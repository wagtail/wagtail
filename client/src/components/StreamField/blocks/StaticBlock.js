import { escapeHtml as h } from '../../../utils/text';

export class StaticBlock {
  constructor(blockDef, placeholder) {
    this.blockDef = blockDef;

    const element = document.createElement('div');

    if (this.blockDef.meta.html) {
      element.innerHTML = this.blockDef.meta.html;
    } else {
      element.innerHTML = h(this.blockDef.meta.text);
    }

    placeholder.replaceWith(element);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
  setState(_state) {}

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
  setError(_errorList) {}

  getState() {
    return null;
  }

  getValue() {
    return null;
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  focus() {}
}

export class StaticBlockDefinition {
  constructor(name, meta) {
    this.name = name;
    this.meta = meta;
  }

  render(placeholder) {
    return new StaticBlock(this, placeholder);
  }
}
