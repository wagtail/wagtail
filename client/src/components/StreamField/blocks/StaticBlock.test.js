/* eslint-disable @typescript-eslint/no-unused-vars */

import { StaticBlockDefinition } from './StaticBlock';

import $ from 'jquery';
window.$ = $;

window.comments = {
  getContentPath: jest.fn(),
};

describe('telepath: wagtail.blocks.StaticBlock', () => {
  let boundBlock;

  beforeEach(() => {
    // Define a test block
    const blockDef = new StaticBlockDefinition(
      'test_field',
      {
        text: 'The admin text',
        icon: 'icon',
        label: 'The label',
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'));
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('telepath: wagtail.blocks.StaticBlock HTML escaping', () => {
  let boundBlock;

  beforeEach(() => {
    window.somethingBad = jest.fn();

    // Define a test block
    const blockDef = new StaticBlockDefinition(
      'test_field',
      {
        text: 'The admin text <script>somethingBad();</script>',
        icon: 'icon',
        label: 'The label',
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'));
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('javascript can\'t execute', () => {
    expect(window.somethingBad.mock.calls.length).toBe(0);
  });
});

describe('telepath: wagtail.blocks.StaticBlock allows safe HTML', () => {
  let boundBlock;

  beforeEach(() => {
    window.somethingBad = jest.fn();

    // Define a test block
    const blockDef = new StaticBlockDefinition(
      'test_field',
      {
        html: 'The admin text <script>somethingBad();</script>',
        icon: 'icon',
        label: 'The label',
      }
    );

    // Render it
    document.body.innerHTML = '<div id="placeholder"></div>';
    boundBlock = blockDef.render($('#placeholder'));
  });

  test('it renders correctly', () => {
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('javascript can execute', () => {
    expect(window.somethingBad.mock.calls.length).toBe(1);
  });
});
