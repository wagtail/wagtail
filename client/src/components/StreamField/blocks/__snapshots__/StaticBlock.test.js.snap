// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telepath: wagtail.blocks.StaticBlock HTML escaping it renders correctly 1`] = `"<div>The admin text &lt;script&gt;somethingBad();&lt;/script&gt;</div>"`;

exports[`telepath: wagtail.blocks.StaticBlock allows safe HTML it renders correctly 1`] = `"<div>The admin text <script>somethingBad();</script></div>"`;

exports[`telepath: wagtail.blocks.StaticBlock it renders correctly 1`] = `"<div>The admin text</div>"`;
